### PR TITLE
Added InitializeObject hook

### PR DESCRIPTION
**Pull request by 256dpi**
Hi there!

First of all thanks for this amazing library!

I'm working on a small very opinionated framework that uses struct tags to simplify the declaration of models. I first thought this might get very complicated with `api2go` as it uses `reflect` to create new objects internally. But in fact I learned that most actions go either through `crud.Find()` or `crud.FindAll()` which allows to setup the object before `jsonapi.Marshal()` or `jsonapi.Unmarshal()` is called. The situation that does not allow such access is on a PATCH request. An object gets created with `reflect` right before passing it to `jsonapi.Unmarshal()`.

TL;DR:
This pull request adds an optional method that can be implemented to get access to the dynamically created object in a PATCH request before `jsonapi.Unmarshal()` is called. This is necessary to give context to the models in higher frameworks that abstract the `jsonapi` interfaces.

Please feel free to change the name of the interface and method. ;)

### CODE REVIEW

**Comment by wwwdata** [comment]
Thanks for the PR. Yes what you are saying makes sense. Im just confused that you mentioned `PATCH`. You mean `POST` because you called the potential interface implementation in the Create api route, right? Which makes totally sense because in that, we create a new instance of a struct and then unmarshal into it, which would be the undesired behaviour you are describing.

I currently don't really understand why go 1.4 fails in travis, but I guess it's some golint issues.

I like the interface name, but please squash the `fix typo` commit into your first, I'll check that the test are working and then we can merge it :)

**Comment by wwwdata** [comment]
https://github.com/golang/lint/issues/198
> Per the README, golint nowadays requires Go 1.5 or later.

ok, hm, let's ditch that old version 1.4 then :D

**Comment by sharpner** [comment]
I would love if you could add a test using the newly introduced behaviour, and additionally shouldn't the interface be something like ```ObjectInitializer``` to keep golangs interface guidelines?

**Comment by 256dpi** [comment]
Thanks for the fast responses.

@wwwdata Ups, yes I mean `POST` of course.
@sharpner You're right, that makes more sense. I will add a test in a moment.

**Comment by wwwdata** [comment]
once #248 is merged you can just rebase on it for the tests to work

**Comment by wwwdata** [comment]
:+1: just one more rebase before merge

**Comment by 256dpi** [comment]
Thanks!
